### DISCUSSION TITLE
[Feature]Integrate whirlpool into Bisq wallet

### DISCUSSION THREAD

**Discussion by Bl5ckj5ck** [bisq-network/bisq#5887] [Ideas]
GitHub Issue: #5867

Hey!👋
Bisq is one of the few platforms aiming at preserving the privacy of the users and where you can buy/sell bitcoins for fiat with no KYC.
It's a great peer to peer exchange platform but it doesn't give users a complete privacy-oriented solution since it is based on a wallet which lacks any privacy measure intended to protect them against on chain tracking techniques.

Indeed Bisq wallet does not offer any method to keep txs unlinked to each other on the blockchain.
Currently the only solution consists on transferring the funds to an external wallet, mixing them through a coinjoin and sending them back to the Bisq wallet.

We can do more in order to guarantee financial privacy.
Integrating whirlpool into Bisq wallet would be a great step forward to make the common ownership analysis more difficult. It would be also easier and faster for Bisq users to keep their tx private in just a single wallet(Bisq).
Thus, this would increase the privacy of both Bisq and Samourai users.

What is coinjoin?
Coinjoin is a technique developed to create a transaction which makes impossible to trace ownership of funds with
any certainty.

Why whirlpool?
Whirlpool is a coinjoin implementation which offers a high degree of anonimity and lets users have full ownership of their funds. It has gained a lot of trust (it has been integrated also into Sparrow wallet, recently 4558 BTC unspent capacity). There are no deterministic links between inputs and outputs, no address reuse, no mixing previously seen coins together and no mix with yourself.

I'm sure that this feature could be very useful for each Bisq user if implemented. I hope it will be applied. We need privacy to be increased on layer one as much as we can by integrating and using the tools available

**Comment by w0000000t** [comment]
This could introduce traders to the concept of coinjoining making the latter more and more used, ultimately facilitating coinjoins themselves as the more mixing parties the better.
We had a discussion about Whirlpool/JoinMarket and you pointed me to an article where JM was "victim" of a reverse engineering tracing attack, but since then I've read an article where JM is analized more in detail ( https://bitcoinmagazine.com/technical/joinmarket-zerolink-and-coinjoin-bitcoin-mixing ) that analyses several attack vectors and countermeasures, so it makes JM surely harder to use, but it is still the cheapest, and when you know what you're doing it shouldn't lose its effectiveness.

Now, integrating coinjoin into Bisq could be a challenge the devs might not be extremely eager to jump on.

**Reply by Bl5ckj5ck** [reply]
> This could introduce traders to the concept of coinjoining making the latter more and more used

Honestly I don't agree. Onboarding users in coinjoin should be done with a private and friendly option. I wouldn't be surprised if a new user gets intimidated and abandon the idea of using it.
Joinmarket is a cool idea on paper but it's actually a nightmare for the average user. The result is that it's less attractive and less private than Whirlpool because users might easily incur in errors and compromise their privacy.

> ...it makes JM surely harder to use, but it is still the cheapest, and when you know what you're doing it shouldn't lose its effectiveness

I personally wouldn't mind to pay more and be sure that my cojnjoin completes safely and privately. Whirlpool could be a bit more expensive but it's an upfront cost which balances with free "unlimited" remixing. Again, users are inclined to errors especially when they are new and if the tool which they use is challenging. Is it worth paying less and running a high risk of compromising your privacy? The tool definitely loses its effectiveness unless you "enforce" privacy steps which prevent users from doing errors
